studygroups/views/facilitate.py

```diff
@@ -546,48 +548,111 @@ def add_member(request, study_group_id):
+            context['team'] = team_from_token
+            return context
+
+        # if there's an invitation, get the team from the invitation
+        invitation = TeamInvitation.objects.filter(
+            email__iexact=self.request.user.email,
+            responded_at__isnull=True).first()
+
+        if invitation:
+            context['team'] = invitation.team
+            return context
+
+        # otherwise use team with matching email
+        team = eligible_team_by_email_domain(self.request.user)
+        context['team'] = team
         return context
 
     def form_valid(self, form):
-        # Update invitation
-        invitation = TeamInvitation.objects.filter(email__iexact=self.request.user.email, responded_at__isnull=True).first()
-        current_membership_qs = TeamMembership.objects.filter(user=self.request.user)
-        if current_membership_qs.filter(role=TeamMembership.ORGANIZER).exists() and self.request.POST['response'] == 'yes':
+        invitation_confirmed = self.request.POST['response'] == 'yes'
+        user = self.request.user
+        default_role = TeamMembership.MEMBER
+        new_membership = None
+        current_membership_qs = TeamMembership.objects.active().filter(user=user)
+
+        # if you're currently the only organizer for a team, you can't join any other teams
+        if invitation_confirmed and current_membership_qs.filter(role=TeamMembership.ORGANIZER).exists():
             messages.warning(self.request, _('You are currently the organizer of another team and cannot join this team.'))
             return http.HttpResponseRedirect(reverse('studygroups_facilitator'))
-        invitation.responded_at = timezone.now()
-        invitation.joined = self.request.POST['response'] == 'yes'
-        invitation.save()
-        if invitation.joined is True:
-            if current_membership_qs.exists():
+
+        # add to team by token
+        token = self.kwargs.get('token')
+        team_from_token = self.get_team_from_token(token)
+        if team_from_token and invitation_confirmed:
+            current_membership_qs.delete()
+            new_membership = TeamMembership.objects.create(team=team_from_token, user=user, role=default_role)
+
+        # add to team by invitation
+        invitation = TeamInvitation.objects.filter(email__iexact=user.email, responded_at__isnull=True).first()
+        if invitation:
+            invitation.responded_at = timezone.now()
+            invitation.joined = invitation_confirmed
+            invitation.save()
+
+            if invitation_confirmed:
                 current_membership_qs.delete()
-            TeamMembership.objects.create(team=invitation.team, user=self.request.user, role=invitation.role)
+                new_membership = TeamMembership.objects.create(team=invitation.team, user=user, role=invitation.role)
+
+        # add to team by matching email
+        eligible_team = eligible_team_by_email_domain(user)
+        if eligible_team:
```

Comment:
@s-kennedy I didn't test, but I suspect there will be a bug if you accept an invitation for a team while your email matches a different team domain. The code at line 624-630 will run, joining the team. Then the code from 634-637 will run, removing you from the team you just joined and adding you to the eligible_team.

Reply:
@s-kennedy I see [`eligible_team_by_email_domain`](https://github.com/p2pu/learning-circles/blob/team_invitation_link/studygroups/models/team.py#L102) checks if the user is already part of a team and returns `None` if so. Given that we're doing most of the team logic in this view, I'm not sure if that is where the check makes the most sense, but at least I understand now why there isn't a bug :)